bookkeeper-server/src/main/java/org/apache/bookkeeper/bookie/BookieShell.java

```diff
@@ -970,48 +973,87 @@ int runCmd(CommandLine cmdLine) throws Exception {
-            runFunctionWithLedgerManagerFactory(bkConf, mFactory -> {
-                try (LedgerManager m = mFactory.newLedgerManager()) {
-                    LedgerRangeIterator iter = m.getLedgerRanges();
-                    if (cmdLine.hasOption("m")) {
-                        List<ReadMetadataCallback> futures = new ArrayList<ReadMetadataCallback>(LIST_BATCH_SIZE);
-                        while (iter.hasNext()) {
-                            LedgerRange r = iter.next();
-                            for (Long lid : r.getLedgers()) {
-                                ReadMetadataCallback cb = new ReadMetadataCallback(lid);
-                                m.readLedgerMetadata(lid, cb);
-                                futures.add(cb);
-                            }
-                            if (futures.size() >= LIST_BATCH_SIZE) {
-                                while (futures.size() > 0) {
-                                    ReadMetadataCallback cb = futures.remove(0);
-                                    printLedgerMetadata(cb);
+            final boolean printMeta = cmdLine.hasOption("m");
+            final String bookieidToBePartOfEnsemble = cmdLine.getOptionValue("bookieid");
+            final BookieSocketAddress bookieAddress = StringUtils.isBlank(bookieidToBePartOfEnsemble) ? null
+                    : new BookieSocketAddress(bookieidToBePartOfEnsemble);
+            ClientConfiguration conf = new ClientConfiguration();
+            conf.addConfiguration(bkConf);
+
+            try (BookKeeperAdmin bkAdmin = new BookKeeperAdmin(conf)) {
+                runFunctionWithLedgerManagerFactory(bkConf, mFactory -> {
+                    try (LedgerManager ledgerManager = mFactory.newLedgerManager()) {
+
+                        final AtomicInteger returnCode = new AtomicInteger(BKException.Code.OK);
+                        final CountDownLatch processDone = new CountDownLatch(1);
+
+                        Processor<Long> ledgerProcessor = new Processor<Long>() {
+                            @Override
+                            public void process(Long ledgerId, VoidCallback cb) {
+                                if (!printMeta && (bookieAddress == null)) {
+                                    printLedgerMetadata(ledgerId, null, false);
+                                    cb.processResult(BKException.Code.OK, null, null);
+                                } else {
+                                    GenericCallback<LedgerMetadata> gencb = new GenericCallback<LedgerMetadata>() {
+                                        @Override
+                                        public void operationComplete(int rc, LedgerMetadata ledgerMetadata) {
+                                            if (rc == BKException.Code.OK) {
+                                                if ((bookieAddress == null)
+                                                        || bkAdmin.areEntriesOfLedgerStoredInTheBookie(ledgerId,
```

Comment:
I think `areEntriesOfLedgerStoredInTheBookie` and `areEntriesOfSegmentStoredInTheBookie` can be static. If you make them static, we don't need to construct another `BookKeeperAdmin` object.